feat: deprecate StepIndicator v1

langz opened the pull request:
This PR removes StepIndicator v1 🗑️

langz commented on packages/dnb-eufemia/src/components/step-indicator/StepIndicatorList.js:
```diff
@@ -53,16 +51,14 @@ export default class StepIndicatorList extends React.PureComponent {
 
-    // deprecated
-    if (!isV1) {
+    if (useParams) {
```
I'm not 100% sure about what's correct to do in this `if` & `else if` statement 🤔

langz replied:
Whenever I tried removing this if sentence(which I thought should be possible), I got the following errors when running tests: `ReferenceError: Cannot access 'useParams' before initialization`

langz commented on packages/dnb-eufemia/src/components/step-indicator/__tests__/StepIndicator.test.js:
```diff
@@ -109,19 +109,6 @@ describe('StepIndicator Sidebar', () => {
 
-describe('StepIndicator in general', () => {
-  it('should support spacing props', () => {
-    render(<Component data={stepIndicatorListData} top="large" />)
```
I removed this test as it seems like it was actually testing StepIndicator v1, as it was missing a `mode` property.

But do we think the StepIndicator(previously known as v2) should support spacing?
As of now, we have space property listed in properties, https://eufemia.dnb.no/uilib/components/step-indicator/properties/

But when trying to actually use the space properties in any of the examples/demos for StepIndicator(v2) it doesn't seem to actually work, like in this example where it has the property bottom without it actually doing anything: https://eufemia.dnb.no/uilib/components/step-indicator/demos/#stepindicator-in-loose-mode

So, should we support spacing or not for StepIndicator(v2)?

tujoworker replied:
Yes, we should. I though I fixed there something not so long ago.  It it was probably only the sidebar.

langz replied:
I've now added more support spacing, and a few more tests.

langz commented:
Should I make the `mode` property optional now?
If so, which value should it default to? `strict` or `loose`?

langz commented on packages/dnb-eufemia/src/components/step-indicator/StepIndicatorList.js:
```diff
@@ -28,22 +28,17 @@ export default class StepIndicatorList extends React.PureComponent {
 
     const params = {
-      // 'aria-label': 'progress', // used before in v1
-      ...listAttributes,
```
I've removed all references to `listAttributes`, as I couldn't really understand what it was doing, and that it only seemed to be used in "v1", https://github.com/dnbexperience/eufemia/pull/1840/files#diff-663b46770651a3155611a8735085b3ff56979b63d3c378cd7f5a7855fd072983L96

tujoworker replied:
We probably should still keep it, taking [this comment](https://github.com/dnbexperience/eufemia/pull/1840#discussion_r1060558302) in account.

langz commented on packages/dnb-eufemia/src/components/step-indicator/StepIndicatorList.js:
```diff
@@ -53,27 +48,21 @@ export default class StepIndicatorList extends React.PureComponent {
       // prettier-ignore
       [useParams]['aria-labelledby'] = combineLabelledBy(
-        listAttributes,
+        {},
```
I replaced `listAttributes` with `{}`, as I removed `listAttributes` as it was always `{}` as far as I could understand.

tujoworker replied:
We collect the "rest" props and send them in to its local context:
```jsx
listAttributes={this.props}
```

then we pick it up from `this.context`.

So if a dev now "needs" for some reason to define an `aria-labelledby` (or we do?) then it will be a part of `listAttributes`.

Next, `combineLabelledBy` will then "combine" them together, because `aria-labelledby` can support several IDs. This way, we do not overwrite them, but rather combine them. We can't send in `{}`.

We should make a test for it. I can make it if you want?

langz replied:
Done :)

langz commented on packages/dnb-design-system-portal/src/docs/uilib/demos/eufemia-demo/form-demo-01.js:
```diff
@@ -61,22 +61,6 @@ const Header = () => (
-        on_change={(e) => {
-          console.log('StepIndicator.on_change', e)
-        }}
-      />
```
Because of the removal of this, I'm not sure if we should update the screenshot of the "form demo" as well.
https://raw.githubusercontent.com/dnbexperience/eufemia-demo/main/misc/screenshot_comparison.png

langz commented on packages/dnb-design-system-portal/src/docs/uilib/demos/eufemia-demo/form-demo-01.js:
```diff
@@ -61,22 +60,6 @@ const Header = () => (
         <Heading size="x-large">Card complaint</Heading>
       </HeaderTitleWrapper>
-
-      <StepIndicator
```
Because of the removal of this, I'm not sure if we should update the screenshot of the "form demo" as well.
https://raw.githubusercontent.com/dnbexperience/eufemia-demo/main/misc/screenshot_comparison.png

tujoworker replied:
hehe, can we change it to tabs instead? And yes, we then need to update it anyway.

langz replied:
Tabs it is 👍

tujoworker commented:
> Should I make the `mode` property optional now? If so, which value should it default to? `strict`, `static` or `loose`?

If its not too late, we may stick with it as it is, because none of these are suited to be the default one, I think. What do you think?

langz commented:
> > Should I make the `mode` property optional now? If so, which value should it default to? `strict`, `static` or `loose`?
> 
> If its not too late, we may stick with it as it is, because none of these are suited to be the default one, I think. What do you think?

That's perfectly fine ✅

tujoworker commented on packages/dnb-design-system-portal/src/docs/uilib/components/step-indicator/info.md:
```diff
@@ -110,6 +110,6 @@ The API has been simplified and some properties have to be changed by the next m
 - Add a property called `sidebar_id` with an unique ID.
 - Optional, place `<StepIndicator.Sidebar sidebar_id="unique-step-indicator" />` in your layout.
-- URL handling is deprecated and will be removed in v10
+- URL handling is deprecated and is removed in v10
```
We can probably remove this sentence now, that it "was" deprecated.

tujoworker commented on packages/dnb-design-system-portal/src/docs/uilib/components/step-indicator/info.md:
```diff
@@ -110,6 +110,6 @@ The API has been simplified and some properties have to be changed by the next m
+- URL handling is deprecated and is removed in v10
 
-The documentation is updated and every property name that will be deprecated is clearly marked with a strike-through line. You will get a console warning during development when you use outdated properties.
+The documentation is updated and every deprecated property name is clearly marked with a strike-through line. You will get a console warning during development when you use outdated properties.
```
We can probably remove this sentence now, that it "was" deprecated.

tujoworker commented on packages/dnb-eufemia/src/components/step-indicator/StepIndicatorContext.js:
```diff
@@ -39,10 +39,7 @@ const filterAttributes = Object.keys(stepIndicatorDefaultProps)
     'innerRef',
     'hasSkeletonData',
     'filterAttributes',
-    'listAttributes',
```
Should `listAttributes` really be removed? I'm not sure by just looking at the PR code.

tujoworker commented on packages/dnb-eufemia/src/components/step-indicator/StepIndicatorList.js:
```diff
@@ -28,22 +28,17 @@ export default class StepIndicatorList extends React.PureComponent {
       className: classnames(
         'dnb-step-indicator',
         createSkeletonClass('font', skeleton),
-        createSpacingClasses(listAttributes),
```
Same here, we use the "rest" props to create the spacing classes.

langz replied:
Changed to using rest now :)

tujoworker reviewed (approved):
Really nice work 💯 had just some comments about the `listAttributes` thing.

langz commented:
I've now added support for `aria-labelledby,` and added tests for it: https://github.com/dnbexperience/eufemia/pull/1840/commits/a9f2045afb674dc37cb097d77b09004f4c5ad4bb
